### DISCUSSION TITLE
Optional alternative crate-naming scheme

### DISCUSSION THREAD

**Discussion by cfredric** [bazelbuild/rules_rust#1008] [Ideas]
Hello!

Google has some Rust in its source repository, and we have realized that rules_rust's current approach to naming crates (i.e., using the target name as the crate name) will not work for us. We need to use a naming scheme that uniquely identifies each first-party crate, and associates it with the entire Bazel label that it was built from; for third-party crates, we can use the crate's actual name (without modification).

We're experimenting with an approach that embeds the Bazel label of a dependency crate in Rust source code, and uses a proc macro to transform the Bazel labels into valid Rust identifiers corresponding to the crates listed in the target's `deps` (following a particular encoding scheme that must be agreed upon between the proc macro and the rules_rust implementation).

An example of this usage is the following:

```rust
import! {
  "//some_project:utils";
  "//some_project:ugly_name" as something_else;
  "//third_party/rust/serde/v1:serde";
  "//third_party/rust/clap/v2:clap" as aliased_clap;
}
```
This would expand to:

```rust
extern crate some_project_COLON_utils as utils;
extern crate some_project_COLON_ugly_name as something_else;
extern crate serde;
extern crate clap as aliased_clap;
```

Note that the two crates imported from the //third_party directory were not renamed, but otherwise, that the whole Bazel label was visible in source, and encoded in the dependency crate's name.

I've written a prototype of this macro (uploaded in [this commit](https://github.com/cfredric/rules_rust/commit/d6196637fe0677f0b5807e096c673014b3906817) - currently broken due to missing deps, but you can play around with it and see the idea), so now I'd like to add the ability to do this renaming in rules_rust (behind a flag), to continue experimenting.

I plan to add a flag to allow a project to opt-in to this new naming scheme soon, but wanted to open up for discussion :)

**Comment by hlopko** [comment]
(disclaimer, I work with Chris, so it won't come as a surprise what I write next :)

I FWIW like this approach - it makes Rust fit a little bit better into bigger monorepos, while staying compatible with OSS workflows depending on crates.io (since those names won't be mangled by the macro). We wanted to share this work with other rules_rust users to get feedback and potentially contribute the macro and a build_setting that would enable this mode in rules_rust. We don't plan to remove the existing mode.

There is some work to be done in the rules (to set the right flags to rustc etc.). If this turns out useful, we may try to talk to rust-analyzer folks about some further goodies like automatically adding crates to the macro.

And a small caveat, we will need some time to stabilize the implementation, it should be considered experimental at the moment. But we do plan to use it internally, so it will reach stability soon.

**Reply by dfreese** [reply]
Can you explain a little more about what sounds like a hard requirement to have:
> a naming scheme that uniquely identifies each first-party crate, and associates it with the entire Bazel label that it was built from.

What's the primary driver of that?  Potential name collisions? Tooling?

**Reply by hlopko** [reply]
Yes :)
* Name collisions (strawperson argument is naming the rust_library 'utils' in many instances in the monorepo)
* Automatic update of the BUILD file based on the content of the .rs file (so the tooling knows exactly which target to add into the deps)
* Large scale refactoring tooling (a tool can figure out a target from which a symbol comes from from the information in the source file - it doesn't have to look into the BUILD file)

I'm probably forgetting some other reasons, am I @cfredric?

**Reply by cfredric** [reply]
I think that covers it! Facilitating automated BUILD file updates (when a new crate is pulled in to the .rs source file), avoiding name collisions (without requiring the use of `aliases` at each downstream target), and facilitating large scale refactoring (e.g. renaming a `utils` crate would otherwise require looking at the BUILD files to ignore the "false positives" from other crates also named `utils`) are the primary motivations.

**Comment by UebelAndre** [comment]
If https://github.com/bazelbuild/rules_rust/issues/1013 is the tracking issue, could it be a bit more descriptive? 😅 🙏

**Reply by cfredric** [reply]
Sure! I was thinking that this discussion could be the source of truth, but I'm happy to add more detail there.

**Reply by UebelAndre** [reply]
Thanks! It doesn't have to be detailed but `rename_1p_crates flag` feels pretty ambiguous to me 😅

**Reply by cfredric** [reply]
(done - added this description there, with some rephrasing and a bit more detail)

**Comment by dfreese** [comment]
Haven't thought through it entirely, but relying on just a particular file seems to be somewhat fragile.  What happens for tooling in the multi file rust_library case?  Would a user need to specify these macro blocks in each file for them to work?

It also makes rust look quite different between inside and outside of google. Copybara-ing is always a nightmare, but the distinction between first and third party and the macro seem like it would make that quite difficult.

**Reply by cfredric** [reply]
Re: whether multi-file crates would need to repeat this macro in every file, no, I don't believe so, since the macro introduces the deps crates under aliases (except in the cases where they're not renamed at all). So each file in the multi-file crate can refer to the dep via the alias without issue, after the crate root invokes the macro. This works in the prototyping I've done so far.

Re: inside vs. outside Google & copybara, I think there are a few questions we can ask:

1. Could this be the best approach for *some* rules_rust user?
    i. If yes, is this something that rules_rust is interested in supporting?
1. Is this the best solution for Google? (Clearly, a "yes" here implies "yes" to question 1. But 1i is technically still up for debate.)

I read your comment as mostly asking question 2. I wrote a (non-public) design doc exploring this space and arguing that this is the approach that makes the best tradeoffs for Google's use cases; I welcome you to read it if you're interested, and we can discuss there. Feel free to ping me (cfredric@) or hlopko@ for the doc.

And please let me know if I misunderstood!

**Reply by hlopko** [reply]
@dfreese I'm not sure I follow "relying on just a particular file", could you elaborate for me there? Risking that I'll answer a different question I'll try saying that multi-file crates work - the import macro will deal with crate imports, not with imports of submodules of crates. So following is expected usage pattern:

```
import! {
  "//some_project:utils";
}

use utils::UtilsTrait;
```

import macro can also be used from multiple submodules of the current crate.

> It also makes rust look quite different between inside and outside of google

Yes, it's a question of tradeoffs, and it seems to us right now that for first party code this is the better approach.

> Copybara-ing is always a nightmare, but the distinction between first and third party and the macro seem like it would make that quite difficult.

Well all the code using copybara will be in `//third_party`, and there cannot be a dependency from `//third_party` outside of it. So all the code that is open sourced will be using the standard Rust crate naming, and there won't be a need to transform anything.

**Reply by dfreese** [reply]
So I'm less concerned about the approach from a technical perspective, and more from a user perspective.  It now creates two points of reference for importing crates, `lib.rs` and `BUILD`.  For example If you have three `.rs` files (`lib.rs`, `utils.rs`, `process.rs`) in a library.  Does the macro need to be copied into all three of those files?  Are you expecting this to be an, "include-what-you-use," sort of structure where if you use a crate in a particular file, then you have to include it in that macro, even if it's duplicated in `lib.rs`, for example?

The main concern is that it creates two different points of context that you need to understand what's happening in the rust_library target.

**Reply by cfredric** [reply]
> For example If you have three .rs files (lib.rs, utils.rs, process.rs) in a library. Does the macro need to be copied into all three of those files?

No, it only needs to be written in the crate root.

> Are you expecting this to be an, "include-what-you-use," sort of structure where if you use a crate in a particular file, then you have to include it in that macro, even if it's duplicated in lib.rs, for example?

It does behave sort of like an "include-what-you-use" system in the sense that if the crate does not use the macro to bring in a dependency, then it cannot refer to that dependency. But that's not the purpose of this; Rust doesn't do transitive/textual includes (like C++ does, for example), so IWYU semantics are already enforced by the compiler. The primary goal of this proposal is to allow automatic updates of the BUILD file based on the contents of the Rust source file (crate root).

> The main concern is that it creates two different points of context that you need to understand what's happening in the rust_library target.

It's true that this introduces a requirement that the dependency information be duplicated between the BUILD file and the Rust source itself. We think that's ok, because:
* In a large monorepo (which this feature is intended for), some automated BUILD file management system likely exists. The Bazel labels listed in the macro invocation uniquely identify the location of the dep in the repo, so it is straightforward for that tooling to add/remove deps for the appropriate build target.
* We expect the user to spend more time editing/viewing the Rust source file than the BUILD file, so using the Rust source as the "source of truth" for dependency information makes sense.
* This feature is opt-in, so projects for which this tradeoff doesn't make sense are not obligated to deal with the duplication.

**Reply by dfreese** [reply]
> We expect the user to spend more time editing/viewing the Rust source file than the BUILD file, so using the Rust source as the "source of truth" for dependency information makes sense.

True, but for those familiar with cargo, it does break the equivalency between `Cargo.toml` and `BUILD`.

**Reply by davidbarsky** [reply]
> I wrote a (non-public) design doc exploring this space and arguing that this is the approach that makes the best tradeoffs for Google's use cases; I welcome you to read it if you're interested, and we can discuss there.

Hi @cfredric! I'm looking at tackling the same problem with [buck2](https://github.com/facebook/buck2), and while I'm not necessarily convinced that the macro y'all landed on is the right approach for Buck2, I'd be interested in reading the doc to understand the problem space and considerations. My email's hi[at]davidbarsky[dot]com.

**Reply by scentini** [reply]
Hey @davidbarsky, we are working on cleaning up @cfredric's doc and will follow up. I'd also be very interested to learn of any ideas you have around the problem.

**Reply by davidbarsky** [reply]
> Hey @davidbarsky, we are working on cleaning up @cfredric's doc and will follow up. I'd also be very interested to learn of any ideas you have around the problem.

@scentini I appreciate it. My understanding is that there is a strong culture at Google around not needing a human in the loop to make a judgement call around refactoring/renaming an import, but fb—my employer—doesn't _seem_ to value that to the same degree as Google does. Instead, I think a higher priority is to hew more closely to standard Rust idioms and ask the user to disambiguate dependencies/imports when necessary through a fuzzy finder-style interface, as the pain points Rust users have identified is "we don't want to manually update a BUCK file after adding a `use` statement".

(I also don't think I can get full adoption of an `import!`-style macro this late in Rust's adoption curve at fb. I'm _guessing_ that a non-trivial portion of the value proposition of `import!` is that _everyone_ uses it, and that value is significantly diminished if that isn't the case. That is one of the reasons I asked about the document that y'all wrote—I wanted to understand the cultural context/motivations as why y'all considered this to be best choice.)

**Reply by gribozavr** [reply]
@davidbarsky One major reason for us to use an `import!`-style macro at Google is to ensure absence of crate name collisions in a monorepo by generating unique names for Rust crates based on the Bazel package and target name. That is, the crate name of `//a/b/c:util` isn't `util`, it is something like `a_slash_b_slash_c_colon_util`.

**Reply by davidbarsky** [reply]
> @davidbarsky One major reason for us to use an `import!`-style macro at Google is to ensure absence of crate name collisions in a monorepo by generating unique names for Rust crates based on the Bazel package and target name. That is, the crate name of `//a/b/c:util` isn't `util`, it is something like `a_slash_b_slash_c_colon_util`.

Yeah, I can imagine that requiring users to write `a_slash_b_slash_c_colon_util` would be a crummy experience. I'm personally okay with crate name collisions—after all, there's plenty in `fbsource`—hence why I brought up a fuzzy finder and having the user disambiguate in case of conflicts.

Does Google lint against the usage of plain `use` statements in non-vendored code?

**Reply by gribozavr** [reply]
> I'm personally okay with crate name collisions—after all, there's plenty in `fbsource`

What do you expect the users to do when they need to import two crates with colliding names? Rename one of them locally? But that requires putting that renaming mapping into the build system file, right? It can't be all contained in the source file, I think. Wouldn't that go against your goal of having the users edit the BUCK files less?

> Does Google lint against the usage of plain use statements in non-vendored code?

We don't plan to have extra lints. Users are not supposed to spell the unique (mangled) crate name in the source code, but so far we don't feel like there is a need to protect against it because it looks very wrong.

When the `import!` macro expands to a `use` statement that renames the crate locally to a short name.

For example:

```
import! {
  "//gshoe:utils";
  "//gshoe:grody_name" as something_else;
  "//third_party/rust/serde/v1:serde";
  "//third_party/rust/clap/v2:clap" as aliased_clap;
}
use std::collections::HashMap as HashMap;
```

expands to:

```
extern crate gshoe_colon_utils as utils;
extern crate gshoe_colon_grody_name as something_else;
extern crate serde;
extern crate clap as aliased_clap;
use std::collections::HashMap as HashMap;
```

> My understanding is that there is a strong culture at Google around not needing a human in the loop to make a judgement call around refactoring/renaming an import

Going back to this. Refactoring/renaming an import when we are starting with a piece of code that already compiles isn't that bad. All ambiguities are already resolved, so a tool that performs the renaming has all the necessary information available.

On the other hand, adding a new import to a source file is a use case where our `import!` macro makes a difference. When one adds a new dependency to a source file, one must add an `extern crate` declaration and update the dependencies list in the `BUILD`. We want to:
- automate dependency management in `BUILD` files as much as possible,
- have one source of truth for the dependency information.

For example, in C++ this duplication is necessary and we know from experience that it is quite painful (the source file spells the header file name, and the `BUILD` file spells out the Bazel target name that provides that header).

We already have tooling that updates the dependencies in the `BUILD` file based on the import statements in source code (in a language-dependent way, so it understands `#include` in C++, `import` in Java etc.) How does it work? Mostly well, it is not 100% automatic. Since one C++ header can be a member of multiple Bazel targets, it is not always possible to uniquely identify the corresponding Bazel build target that needs to be added. When doing day to day development it isn't a huge problem, but the cost of exceptions adds up and it is annoying at scale for the infrastructure teams.

Thus, it would be ideal for us if the "import" statement in Rust identified the build dependency in a unique way. Then, the dependencies in `BUILD` files can be managed by tooling completely automatically - they could be generated from the "import" statements in the source file.

> Instead, I think a higher priority is to hew more closely to standard Rust idioms and ask the user to disambiguate dependencies/imports when necessary through a fuzzy finder-style interface, as the pain points Rust users have identified is "we don't want to manually update a BUCK file after adding a use statement".

You're saying that your users have the same pain point, but you are arriving at a different solution, prioritizing idiomatic code over 100% automation. That's a tradeoff that you can certainly make, but there are costs to it.

Based on my experience, I'd guess that your users will be satisfied with a fuzzy finder that works 99% of the time. However, the 1% of exceptions multiplied by a huge codebase size will be a huge toil for   the infrastructure teams that run large scale changes throughout the codebase. That is, the large scale change tooling (for renaming or moving symbols, merging or splitting crates etc.) will need to include the complexity to resolve these ambiguities (even having to punt to a human in some cases). That means that a 99% solution is as good as a 0% solution for the infra teams. The only way to avoid the extra complexity in the tooling is to make the mapping between imports, Bazel targets, and filesystem paths completely predictable without exceptions.

**Reply by davidbarsky** [reply]
Thanks for your in-depth response. I really appreciate it as it gives me a valuable insight into the thinking behind this approach.

> What do you expect the users to do when they need to import two crates with colliding names? Rename one of them locally? But that requires putting that renaming mapping into the build system file, right? It can't be all contained in the source file, I think.

I think a user will be required to rename one of the conflicting crates in the BUCK file, but to be entirely honest, we haven't _had_ many conflicts crop up since there's no commonly used crate called `util` or whatever (really, the biggest issue are Thrift-generated targets, but even so, it's something we can fix).

> Wouldn't that go against your goal of having the users edit the BUCK files less?

Not necessarily. If a user needs to edit a `BUCK` file to rename a crate, that is still a meaningful improvement over needing to edit a `BUCK` file every time they want to change their dependencies. I'm not sure if getting to a state of "no manual editing of a `BUCK` file, ever" is a goal that I can reasonably take on. "Don't interrupt the user's flow by making them switch files to add a dependency" is a far more achievable goal, in my view.

> We don't plan to have extra lints. Users are not supposed to spell the unique (mangled) crate name in the source code, but so far we don't feel like there is a need to protect against it because it looks very wrong.

Gotcha—my the question behind my question is "what happens if a user writes `use tracing::error`" inside, e.g., a function body, but based on the procedural macro expansion you showed, the crate `tracing` doesn't _exist_ under that name unless the `import!` macro is written.

> When one adds a new dependency to a source file, one must add an extern crate declaration and update the dependencies list in the `BUILD`.

I apologize for being overly pedantic, but writing `extern crate` is no longer required in the 2018 and 2021 editions of Rust. `use` alone is sufficient, but I think my prior comment about needing to write `import!` still holds.

> Thus, it would be ideal for us if the "import" statement in Rust identified the build dependency in a unique way. Then, the dependencies in BUILD files can be managed by tooling completely automatically - they could be generated from the "import" statements in the source file.

Makes sense.

> Based on my experience, I'd guess that your users will be satisfied with a fuzzy finder that works 99% of the time. However, the 1% of exceptions multiplied by a huge codebase size will be a huge toil for the infrastructure teams that run large scale changes throughout the codebase. That is, the large scale change tooling (for renaming or moving symbols, merging or splitting crates etc.) will need to include the complexity to resolve these ambiguities (even having to punt to a human in some cases). That means that a 99% solution is as good as a 0% solution for the infra teams. The only way to avoid the extra complexity in the tooling is to make the mapping between imports, Bazel targets, and filesystem paths completely predictable without exceptions.

Part of the reason it took me a second to respond is that I checked with a few infrastructure folks about this, and it's... not an issue that's cropped up for them. While I absolutely believe it's an issue at Google, I'll dig deeper to verify my understanding since this is sort of a one-way door.

One more question: does Google carry patches on rust-analyzer so that the automatic import functionality works with `import!` or is that not really handled?

**Reply by gribozavr** [reply]
> One more question: does Google carry patches on rust-analyzer so that the automatic import functionality works with `import!` or is that not really handled?

We have not started to deploy rust-analyzer in the monorepo yet. I do agree that some customization will be needed, since we would want, for example, code completion to complete Bazel target names inside the `import!` block.

**Reply by davidbarsky** [reply]
> We have not started to deploy rust-analyzer in the monorepo yet. I do agree that some customization will be needed, since we would want, for example, code completion to complete Bazel target names inside the `import!` block

At the risk of becoming a backseat driver, I believe this functionality _might_ need to exist outside of rust-analyzer itself since it'd require  rust-analyzer to be aware of all Rust targets inside of the monorepo. In any case, we've had rust-analyzer deployed for users for several years now and I'd be happy to talk through what we did, what worked, and what didn't.

I know the teams responsible for clangd at our respective employers chat regularly and I've been meaning to set up something regular for Rust as well. Feel free to send me an email at the email I've provided in a previous comment. I think we will face many of the same problems and land on similar solutions and I think it'd be productive to chat and get alignment.

**Reply by gribozavr** [reply]
Of course, rust-analyzer can't be by itself aware of all targets in the monorepo. But I don't think that means that this feature shouldn't be in rust-analyzer. I think it means that for rust-analyzer to work in a large monorepo it needs to be adjusted to not assume anymore that it can parse the whole repo, and instead use an API to access a prebuilt index. For example, see [`class SymbolIndex`](https://github.com/llvm/llvm-project/blob/main/clang-tools-extra/clangd/index/Index.h) in clangd.

**Reply by davidbarsky** [reply]
Yeah, I'm decently familiar with clangd's remote index. I think https://github.com/rust-lang/rust-analyzer/issues/4712 is the equivalent issue on rust-analyzer's end, with the most complete summary being [this comment](https://github.com/rust-lang/rust-analyzer/issues/4712#issuecomment-899411567).

I was thinking on-and-off about how to support that feature properly, given that the people who'd benefit most from this are people with systems like [Glean](https://glean.software/) or [Kythe](https://kythe.io/), but I'm hoping that the approach would be something along the lines of "rust-analyzer can consume some standard-ish format that both Glean and Kythe produce (maybe SCIP)?"
